openbook_auth/models.py

```diff
@@ -3194,6 +3554,61 @@ def _check_is_not_blocked_with_user_with_id(self, user_id):
         if self.is_blocked_with_user_with_id(user_id=user_id):
             raise PermissionDenied(_('This account is blocked.'))
 
+    def _check_can_report_comment_for_post(self, post_comment, post):
+        self._check_has_not_reported_post_comment_with_id(post_comment_id=post_comment.pk)
+        self._check_can_see_post(post=post)
+
+        if post_comment.commenter_id == self.pk:
+            raise ValidationError(
+                _('You cannot report your own comment.'),
+            )
+
+    def _check_has_not_reported_post_comment_with_id(self, post_comment_id):
+        if self.has_reported_post_comment_with_id(post_comment_id=post_comment_id):
+            raise ValidationError(
+                _('You have already reported the comment.'),
+            )
+
+    def _check_can_report_post(self, post):
+        self._check_can_see_post(post=post)
+        self._check_has_not_reported_post_with_id(post_id=post.pk)
+        if post.creator_id == self.pk:
+            raise ValidationError(
+                _('You cannot report your own post.'),
+            )
+
+    def _check_has_not_reported_post_with_id(self, post_id):
+        if self.has_reported_post_with_id(post_id=post_id):
+            raise ValidationError(
+                _('You have already reported the post.'),
+            )
+
+    def _check_can_report_user(self, user):
+        self._check_has_not_reported_user_with_id(user_id=user.pk)
+        if self.pk == user.pk:
+            raise ValidationError(
+                _('You cannot report yourself...'),
+            )
+
+    def _check_has_not_reported_user_with_id(self, user_id):
+        if self.has_reported_user_with_id(user_id=user_id):
```

Comment:
So users can only be reported once? Even though different categories are available, and a user might be violating more than one?

Reply:
Same goes for the other instances, communities, posts etc.

Reply:
There's a description you can add up ur reportz